### ISSUE TITLE
Allow setting https-only on a ​per-application basis

### ISSUE THREAD

**Issue by rimusz** [deis/controller#579]
Controller should allow setting https-only on a ​per-application basis.
That would allow secure controller itself and run Apps with HTTPS/HTTP support

**Comment by helgi** [comment]
Can you elaborate on that? Attaching a certificate and then doing enforceHTTPS on the router then it gets HTTPS only, otherwise you have *per application* TLS

**Comment by krancour** [comment]
@helgi the issue really is that there's no way via the controller to enable enforcement of HTTPS only (i.e. a redirect from HTTP to HTTPS) on an app by app basis.  Currently, all we can accommodate is that the router can enforce this globally.

I had an offline discussion with @rimusz and what he really wants to do is force HTTPS for the controller only, but not for any other apps.  (Remember that the router regards the controller as "just another app.")  On the router end, it would be very easy to start making the determination re: enforcement of HTTPS on an app-by-app basis if the controller can start annotating apps' service resources with such information.  (This would require a complementary CLI change as well.)

With all of that in place, it would be possible to secure the controller only as @rimusz is looking to do-- although in that particular case, it would be achieved by directly modifying the controller's service resource and _not_ through the CLI.  Even so, the crux of the issue here is that enforcing HTTPS is currently possible for all apps or none and @rimusz would like to see that change.

**Comment by bacongobbler** [comment]
How do we want the client to look? I'm having a hard time thinking of a good client endpoint. `deis apps:ssl enable`, maybe? Any ideas?

**Comment by helgi** [comment]
I think that's fine, tho do we envision to ever add more ssl related options?

**Comment by bacongobbler** [comment]
that's the problem. I do believe we would want to add more ssl-related options, but certs currently exist as `deis certs` so `deis ssl` as a top level feature would just look confusing, IMO.

**Comment by slack** [comment]
No strong preference! Another option would be `ssl:require enable`, which would leave the `ssl:*` namespace for things like cipher lists, and... I don't know what else!

**Comment by bacongobbler** [comment]
Indeed. Users would probably want to supply app-specific cipher suites, whitelisted protocols, session timeouts, buffer sizes etc. which are all technically possible thanks to https://github.com/deis/router/pull/237. It's just a matter of plugging that into the nginx config.

**Comment by slack** [comment]
I think `certs` being its own thing is okay, as the lifecycle of certificates are decoupled from the app...

**Comment by bacongobbler** [comment]
So for a more formal proposal with `deis ssl` as a first class citizen:

```
deis ssl:enforce true
deis ssl:protocols "TLSv1 TLSv1.1 TLSv1.2"
deis ssl:ciphers "ECDHE-ECDSA-CHACHA20-POLY1305:..."
```

The more I think about it, the more I'd like this to be part of the Config/Release model rather than the AppSettings model. That way users can roll back changes as required and see what if SSL was enforced this release etc.

**Comment by helgi** [comment]
Lets first answer the question if any of those options require a deploy of the application?

You are on to something there, I think it can be part of the Config model but it also makes me wonder more and more about AppSettings. 

AppSettings are items that could also be part of a release but given how we use versions as part of labels we can't increment those without a real deploy... Puts AppSettings into a weird situation

Basically, did we do ourselves a disservice by using version in the labels?

**Comment by bacongobbler** [comment]
> Lets first answer the question if any of those options require a deploy of the application?

Short answer: no, it does not technically require a new application release in order to affect the application, since it is all router-side configuration. However, they are still significant enough that users will want to roll back to previously set tls config. For example, if I removed a TLS protocol, changed the cipher suite and then enforced TLS, I might want to revert back to before I did all that. Without a way to roll back, I must manually re-configure these back to the old config.

> Basically, did we do ourselves a disservice by using version in the labels?

When did this occur? I don't know enough about this to comment on that change.

**Comment by helgi** [comment]
> When did this occur? I don't know enough about this to comment on that change.

As long as I can remember - A lot of our matching is done that way. Pod A belongs to RC / RS X and the versioning label is the only thing that really separated things

We can do without it since `Deployments` / `RS` use the pod-template-hash (auto generated) to make things unique. Tho it would make it harder for us to identify what RS / Pod belongs to what release, unless we are fine with seeing Pods belonging to release v40 but then latest the user can see it v48 due to them changing bunch of non-deploy settings (like what we are talking about)

We could mark Release objects as `non-deploy` in `deis releases` and document how that all behaves

**Comment by bacongobbler** [comment]
Oh I see what you mean now. I thought you were referring to the router labels we were attaching to the app service. Since the version label only occurs for the deployment/RC it doesn't really affect AppSettings all that much, since AppSettings just modifies the Service at this point. I actually think that's a good thing that we have the release number in the pod label because it gives users a quick glance on what version is running in the cluster.

I think I'll handle `deis tls` the same way as we handle `deis domains` for a first try and see how that goes, then we can figure out if rollbacks are required in the future. Ye Savvy?

**Comment by helgi** [comment]
Ah right, that's an understandable confusion. As far as the "at glance" users have lost some of that since versions are no longer in pod names which is why I was wondering about the ability to say "deis releases" produces a list of `non-deploy` and `deploy` versions so we could merge AppSettings back into Config and have `rollback` behaviour for both.

With a whole new model to manage it?

**Comment by bacongobbler** [comment]
> With a whole new model to manage it?

yarp

**Comment by helgi** [comment]
Hmm - well give it a go if you want. I think it is a bit of an overkill. How would we migrate that to Config / AppSettings in the future? And what does it bring over AppSettings in this case?

**Comment by bacongobbler** [comment]
> How would we migrate that to Config / AppSettings in the future?

After thinking it through we probably won't, same as we handle certs and domains. I'll come up with an implementation and we can argue about its implementation over there ;)

**Comment by bacongobbler** [comment]
Running TODO checklist:

 - [x] workflow-cli changes (https://github.com/deis/workflow-cli/pull/191)
 - [x] controller-sdk-go changes (https://github.com/deis/controller-sdk-go/pull/82)
 - [x] controller changes (https://github.com/deis/controller/pull/1004)
 - [ ] documentation changes
 - [ ] e2e changes

**Comment by felixbuenemann** [comment]
@bacongobbler Keep in mind that most nginx ssl settings cannot be changed per sni vhost (to be precise they can be set, but have no effect), so to really have app specific TLS settings you'd have to use different listen ports or ips per app.

**Comment by helgi** [comment]
Can you link to the documentation for that? I ask because I've not had any problems with SNI + multiple apps on a single IP in the past. When I've seen SNI and multiple IPs being needed is XP-era browsers and non-SNI compliant libraries.

**Comment by felixbuenemann** [comment]
Yes, that works, but you can't alter most of the ssl setting between sni vhosts that share the same ip and port, so eg. changing cipher suites or protocols per app wouldn't work.
